File: solution/0000-0099/0020.Valid Parentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (stack.isEmpty() || !match(stack.pop(), c)) {
                return false;
            }
        }
        return stack.isEmpty();
    }

    private boolean match(char l, char r) {
        return (l == '(' && r == ')') ||
            (l == '{' && r == '}') ||
            (l == '[' && r == ']');
    }
}

File: solution/0000-0099/0020.Valid Parentheses/Solution.go
func isValid(s string) bool {
    stack := []rune{}
    for _, c := range s {
        if c == '(' || c == '{' || c == '[' {
            stack = append(stack, c)
        } else if len(stack) == 0 || !match(stack[len(stack)-1], c) {
            return false
        } else {
            stack = stack[:len(stack)-1]
        }
    }
    return len(stack) == 0
}

func match(l, r rune) bool {
    return (l == '(' && r == ')') || (l == '{' && r == '}') || (l == '[' && r == ']');
}
